go/codegen/alps/template_train.go

```diff
@@ -0,0 +1,168 @@
+// limitations under the License.
+
+package alps
+
+import "sqlflow.org/sqlflow/go/ir"
+
+type trainFiller struct {
+	DataSource        string
+	TrainSelect       string
+	ValidationSelect  string
+	Estimator         string
+	FieldDescs        map[string][]*ir.FieldDesc
+	FeatureColumnCode string
+	Y                 *ir.FieldDesc
+	ModelParams       map[string]interface{}
+	TrainParams       map[string]interface{}
+	ValidationParams  map[string]interface{}
+	Save              string
+	TmpTrainTable     string
+	TmpValidateTable  string
+}
+
+var templateTrain = `import copy
+import os
+import shutil
+
+import tensorflow as tf
+from alps.framework.column.column import (DenseColumn, GroupedSparseColumn,
+                                          SparseColumn)
+from alps.framework.engine import LocalEngine
+from alps.framework.experiment import EstimatorBuilder
+from alps.io.base import OdpsConf
+from runtime import db
+from runtime.alps.train import train
+from runtime.tensorflow.get_tf_version import tf_is_version2
+
+feature_column_names = [{{range $target, $desclist := .FieldDescs}}{{range $desclist}}
+"{{.Name}}",
+{{end}}{{end}}]
+
+feature_metas = dict()
+{{ range $target, $desclist := .FieldDescs }}
+{{ range $value := $desclist }}
+feature_metas["{{$value.Name}}"] = {
+    "feature_name": "{{$value.Name}}",
+    "dtype": "{{$value.DType | DTypeToString}}",
+    "delimiter": "{{$value.Delimiter}}",
+    "format": "{{$value.Format}}",
+    "shape": {{$value.Shape | intArrayToJSONString}},
+    "is_sparse": "{{$value.IsSparse}}" == "true"
+}
+{{end}}
+{{end}}
+
+label_meta = {
+    "feature_name": "{{.Y.Name}}",
+    "dtype": "{{.Y.DType | DTypeToString}}",
+    "delimiter": "{{.Y.Delimiter}}",
+    "shape": {{.Y.Shape | intArrayToJSONString}},
+    "is_sparse": "{{.Y.IsSparse}}" == "true"
+}
+
+model_params=dict()
+{{range $k, $v := .ModelParams}}
+model_params["{{$k}}"]={{$v | attrToPythonValue}}
+{{end}}
+
+# Construct optimizer objects to pass to model initializer.
+# The original model_params is serializable (do not have tf.xxx objects).
+model_params_constructed = copy.deepcopy(model_params)
+for optimizer_arg in ["optimizer", "dnn_optimizer", "linear_optimizer"]:
+    if optimizer_arg in model_params_constructed:
+        model_params_constructed[optimizer_arg] = eval(model_params_constructed[optimizer_arg])
+
+if "loss" in model_params_constructed:
+    model_params_constructed["loss"] = eval(model_params_constructed["loss"])
+
+
+class SQLFlowEstimatorBuilder(EstimatorBuilder):
+    def _build(self, experiment, run_config):
+        feature_columns_map = {{.FeatureColumnCode}}
+        if feature_columns_map.get("feature_columns"):
+            feature_columns = feature_columns_map["feature_columns"]
+        else:
+            raise ValueError("Not supported feature column map")
+        model_params_constructed["feature_columns"] = feature_columns
+        return tf.estimator.{{.Estimator}}(config=run_config,
```

Comment:
Seems that the alps submitter only supports the builtin TensorFlow estimator? And does it not support feature column except `DENSE` and `SPARSE`?

Reply:
1. Custom estimators can also be supported I think.
2. ALPS supports dense features, sparse features and grouped sparse features, all these type of feature inputs will be supported and add to unit test in incoming pull requests.